src/st_carousel.c

```diff
@@ -0,0 +1,159 @@
+//
+// Copyright(C) 2024 Roman Fomin
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#include "d_player.h"
+#include "doomdef.h"
+#include "doomstat.h"
+#include "doomtype.h"
+#include "g_game.h"
+#include "m_array.h"
+#include "m_misc.h"
+#include "r_defs.h"
+#include "st_sbardef.h"
+#include "v_fmt.h"
+#include "v_video.h"
+
+static const char *names[] = {
+    [wp_fist] = "SMFIST",
+    [wp_pistol] = "SMPISG",
+    [wp_shotgun] = "SMSHOT",
+    [wp_chaingun] = "SMMGUN",
+    [wp_missile] = "SMLAUN",
+    [wp_plasma] = "SMPLAS",
+    [wp_bfg] = "SMBFGG",
+    [wp_chainsaw] = "SMCSAW",
+    [wp_supershotgun] = "SMSGN2",
+};
+
+static const weapontype_t weapon_order[] = {
+    wp_fist,
+    wp_chainsaw,
+    wp_pistol,
+    wp_shotgun,
+    wp_supershotgun,
+    wp_chaingun,
+    wp_missile,
+    wp_plasma,
+    wp_bfg,
+};
+
+typedef struct
+{
+    weapontype_t weapon;
+    int state;
+    boolean darkened;
+} weapon_icon_t;
+
+static weapon_icon_t *weapon_icons;
+
+static weapontype_t curr;
+static int curr_pos;
+
+static int duration;
+
+void ST_UpdateCarousel(player_t *player)
+{
+    if (G_Carousel())
+    {
+        duration = TICRATE * 2;
```

Comment:
Some weapon mods have fast weapon switching (e.g. xa-vesper) so this feels too long in those cases. Try the changes below. The idea is to have the icons appear for 500 ms after the weapon stops moving upward, regardless of the weapon switching time.

```diff
     if (G_Carousel())
     {
-        duration = TICRATE * 2;
+        duration = TICRATE / 2;
     }
 
     if (duration == 0)
     {
         return;
     }
 
-    --duration;
+    if (player->switching == weapswitch_none)
+    {
+        --duration;
+    }
```

Reply:
May need a `if (menuactive) duration = 0;` somewhere too.

Reply:
Done in c0f9b6e